### ISSUE TITLE
Dag Convolution

### ISSUE THREAD

**Issue by ArchieGertsman** [pyg-team/pytorch_geometric#7346]
### 🚀 The feature, motivation and pitch

The partial ordering inherent to dags (directed acyclic graphs) can be incorporated into learning by passing messages one level of the dag at a time, instead of globally. I am proposing a convolution layer that applies a sub-convolution to one level of a dag at a time. A minimal example:

```Python
class DagConv(nn.Module):
    def __init__(self, conv):
        super().__init__()
        self.conv = conv
        
    def forward(self, x, edge_index):
        edge_masks = construct_message_passing_masks(edge_index, x.shape[0])
        
        if self.conv.flow == 'target_to_source':
            edge_masks = reversed(edge_masks)
            
        # message passing one level at a time
        for edge_mask in edge_masks:
            x = self.conv(x, edge_index[:, edge_mask])

        return x
```

where `construct_message_passing_masks` is a subroutine that constructs a list of edge masks, one for each level of the dag, indicating which edges are present at that level. It can be implemented by creating a `NetworkX` graph from `edge_idex` and then calling `nx.topological_generations`.

`DagConv` can be general by allowing the user to specify which convolution to apply at each message passing step. For example, to compute the critical path starting from each node:
```Python
class CriticalPathConv(MessagePassing):
    def __init__(self):
        super().__init__(aggr='max', flow='target_to_source')
        
    def forward(self, x, edge_index):
        return x + self.propagate(edge_index, x=x)


edge_index = torch.tensor([
    [0, 0, 1, 2],
    [1, 2, 3, 3]
])
x = torch.tensor([1, 3, 0, 2]).unsqueeze(-1)

dag_conv = DagConv(CriticalPathConv())
y = dag_conv(x, edge_index).squeeze()
assert y.equal(torch.tensor([6, 5, 2, 2]))
```

### Alternatives

I am not aware of any alternative ways of processing dags with PyG.

### Additional context

This feature request is inspired by [this discussion](https://github.com/pyg-team/pytorch_geometric/discussions/7171). There is existing literature about learning over dags, e.g. DAGNN, whose code can be found at [this repo](https://github.com/vthost/DAGNN) (specifically [this folder](https://github.com/vthost/DAGNN/tree/main/ogbg-code/model)). Their implementation appears to be specific to the architectures tested in the paper, which are a bit complex and overkill for e.g. my application.

**Comment by rusty1s** [comment]
I think your idea is definitely the way to go. Are you asking here to integrate such functionality into PyG? I feel that this can be implemented well already outside `MessagePassing` (as you also did), and `nx.topological_generations` is probably best implemented as a pre-processing step, so I am not yet sure how a PyG integration (besides a desperately needed example) would look like.

**Comment by ArchieGertsman** [comment]
If a complete example is the most helpful, then I'm interested in contributing! I am already doing this in the context of RL-based dag scheduling, (see [architecture](https://github.com/ArchieGertsman/spark-sched-sim/blob/f0d352eb59fbbeb5d23d6767ba5df2cf2395142d/spark_sched_sim/schedulers/decima.py#L266-L360), [nx utility function](https://github.com/ArchieGertsman/spark-sched-sim/blob/f0d352eb59fbbeb5d23d6767ba5df2cf2395142d/spark_sched_sim/graph_utils.py#L169-L194)) but all the other RL code makes for a cumbersome example (environment, training algorithm, etc.). I suppose the environment could be packaged away and the training code could be stripped down, but I'm not sure if RL examples are welcome here. Maybe a minimal re-implementation of DAGNN would be more appropriate.

**Comment by ArchieGertsman** [comment]
See PR #7487
